test/test_type_promotion.py

```diff
@@ -402,6 +402,39 @@ def _test_spot(a, b, res_dtype):
                    torch.tensor(1., dtype=torch.complex64, device=device), torch.complex128)
         _test_spot(torch.tensor([1, 1], dtype=torch.bool, device=device), 1., torch.get_default_dtype())
 
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_result_type_variadic(self, device):
+        # Floating dtypes are not listed here because the `numpy.result_type` may behave differently
+        # if inputs of mixed dtypes are given. The tests for floating type inputs are left to
+        # `test_result_type` and `test_result_type_tensor_vs_scalar`.
+        dtypes = [torch.bool, torch.uint8, torch.int16, torch.int64]
+        inputs = [
```

Comment:
Instead of comparing with NumPy what about implementing the reference using the addition operator?

```
def result_type_ref(args):
  assert len(args) > 0

  t = torch.tensor(True)
  for arg in args:
    if isinstance(arg, torch.dtype):
      # See comment below for a discussion of what type of object dtype should be represented by
      t = t + torch.tensor((1,), dtype=arg)
    else: 
      t = t + arg
  return t.dtype
```

I think this would allow for a simpler test structure and more thorough testing, plus it would compare with the ground truth of what PyTorch is doing

Reply:
I think the reference implementation using `+` sometimes returns unexpected results.
```py
>>> a = torch.tensor([1], dtype=torch.int8)
>>> b = 1.
>>> c = torch.tensor([1], dtype=torch.float16)
>>> (a + b + c).dtype
torch.float32
>>> (a + c + b).dtype
torch.float16
```
Do you have some ideas for this issue?

Reply:
That's really cool!

And it's true, the where our dtypes work the addition operator isn't commutative. Both JAX and NumPy's result_type appears to handle this case correctly.

So I think my suggestion was mistaken. Sorry about that @asi1024. I didn't realize that result_type() can't be thought of as a series of binary elementwise operations. You are right and we can't use this as a reference implementation. 

Returning to the previous tests could be a reasonable solution. If we want to extend that further we could add some "golden value" tests with manually generated checks.

Reply:
After I thought further, I have an idea for a reference implementation.

```py
def result_type_ref(args):
  assert len(args) > 0
  tensor = torch.tensor([True])
  tensor_0dim = torch.tensor(True)
  scalar = True

  for arg in args:
    if isinstance(arg, torch.dtype):
      tensor = tensor + torch.tensor([1], dtype=arg)
    elif isinstance(arg, torch.Tensor):
      if arg.ndim == 0:
        tensor_0dim = tensor_0dim + arg
      else:
        tensor = tensor + arg
    elif isinstance(arg, Number):
      scalar = scalar + arg
    else:
      assert False, "unknown type"
  return ((tensor + tensor_0dim) + scalar).dtype
```

It just uses the same logic as the result_type implementation in this PR. Can I use this implementation in my tests?

Reply:
I think you're right, that's a clever approach, @asi1024. Let's try using this